### ISSUE TITLE
Support for JSX 3 and [@react.component]

### ISSUE THREAD

**Issue by tizoc** [reasonml-community/reductive#44]
Was playing with Reactive and noticed that Provider components don't work with components defined using the new JSX 3 syntax.

Is support for that planned?

I'm not sure whats the proper way (if there is one) for supporting both new style JSX3 components and old-style JSX2 components.

**Comment by nlfiedler** [comment]
I would love to fix this, if I knew how to either 1) support JSX3 or 2) support both. But I cannot find any reference material discussing the changes with JSX3. Can anyone offer some references, please?

**Comment by tizoc** [comment]
I'm not using reactive, but the version I wrote for playing with it using JSX3 looked like this:

```reason
module Lense = {
  type state('reductiveState) = {
    reductiveState: option('reductiveState),
    unsubscribe: option(unit => unit),
  };

  type action =
    | UpdateState
    | AddListener(action => unit);

  module type CONFIG = {
    type state;
    type lensed;
    type action;
    let store: Reductive.Store.t(action, state);
    let lense: state => lensed;
  };

  module Make = (Config: CONFIG) => {
    [@react.component]
    let make =
        (
          ~component:
             (~state: Config.lensed, ~dispatch: Config.action => unit) =>
             React.element,
        ) => {
      let reducer = (state, action) =>
        switch (action) {
        | AddListener(send) => {
            unsubscribe:
              Some(
                Reductive.Store.subscribe(Config.store, _ =>
                  send(UpdateState)
                ),
              ),
            reductiveState:
              Some(Config.lense(Reductive.Store.getState(Config.store))),
          }
        | UpdateState =>
          Config.lense(Reductive.Store.getState(Config.store))
          ->Some
          ->(Belt.Option.eq(state.reductiveState, (a, b) => a === b))
            ? state
            : {
              ...state,
              reductiveState:
                Some(Config.lense(Reductive.Store.getState(Config.store))),
            }
        };

      let initialState = {
        reductiveState:
          Some(Config.lense(Reductive.Store.getState(Config.store))),
        unsubscribe: None,
      };

      let (state, dispatch) = React.useReducer(reducer, initialState);

      React.useEffect1(
        () => {
          dispatch(AddListener(dispatch));
          Some(
            () =>
              switch (state.unsubscribe) {
              | Some(unsubscribe) => unsubscribe()
              | None => ()
              },
          );
        },
        [|state|],
      );

      switch (state.reductiveState) {
      | None => ReasonReact.null
      | Some(state) =>
        component(~state, ~dispatch=Reductive.Store.dispatch(Config.store))
      };
    };
  };
};

module Provider = {
  type state('reductiveState) = Lense.state('reductiveState);
  type action = Lense.action;

  module type CONFIG = {
    type state;
    type action;
    let store: Reductive.Store.t(action, state);
  };

  module Make = (Config: CONFIG) => {
    include Lense.Make({
      include Config;
      type lensed = state;
      let lense = state => state;
    });
  };
};
```

**Comment by tizoc** [comment]
It uses a functor instead of a `createMake` function that returns a function like the current implementation.

Example:

```reason
module MyProvider = Reductive.Provider.Make({
  type state = stateType;
  type action = actionType;
  let store = makeStore();
});
```

and

```reason
module MyProviderWithLense = Reductive.Lense.Make({
  type state = stateType;
  type lensed = lensedType;
  type action = actionType;
  let store = makeStore();
  let lense = state => state.something;
});

```

**Comment by MargaretKrutikova** [comment]
@tizoc , nice implementation! Another way would be to not use lenses at all and go with hooks, like they now have in `react-redux`. The api becomes much simpler in that case, you would do something like:
```
let counter = useSelector(state => state.counter);
```
without the need to wrap components into lenses. 
This can be implemented by putting the store into `React` context, exposing `Provider` and using `useContext` to implement `useSelector`.
I have been playing with my own implementation a bit, so if anyone is interested I can post some code or make a PR 🙂
However, I assume it is not going to support `JSX2` syntax for components.

**Comment by tizoc** [comment]
@MargaretKrutikova a hooks API would be much better, and having React manage the state too. What I couldn't figure out was how to subscribe to a sub-part (the `counter` property in your example) of the context without triggering a re-render of all the components that only care about a small part when the root context changes.

I am interested in your implementation, and more so if you were able to solve the above issue.

**Comment by MargaretKrutikova** [comment]
Yes, I solved the issue with unnecessary rerenders, but I took a lot of inspiration from `react-redux` and studied the code on their side.
I created a `useSelector` hook which selects the necessary state portion and keeps it in a reference (`useRef`).  It subscribes to the store and every time some action is dispatched it re-queries the state and compares it to the state stored in the ref (comparison by reference). If the state has changed, it forces a rerender by incrementing a dummy property on state (it literally does it [here](https://github.com/reduxjs/react-redux/blob/master/src/hooks/useSelector.js#L102) 🙂). So the whole implementation of the hook looks like this:
```reason
let useSelector = selector => {
  let storeFromContext = React.useContext(StoreContext.context);
  let (_, forceUpdate) = React.useReducer((s, _) => s + 1, 0);

  let latestSelectedModel =
    React.useRef(selector(Reductive.Store.getState(store)));

  React.useLayoutEffect1(
    () => {
      let checkForUpdates = () => {
        let newSelectedState = selector(Reductive.Store.getState(store));

        let hasStateChanged =
          newSelectedState !== React.Ref.current(latestSelectedModel);

        if (hasStateChanged) {
          React.Ref.setCurrent(latestSelectedModel, newSelectedState);
          forceUpdate();
        };
      };
      Some(Reductive.Store.subscribe(storeFromContext, checkForUpdates));
    },
    [|storeFromContext|],
  );

  React.Ref.current(latestSelectedModel);
};
```
I skipped some parts with the context, but I have a full working example here on [my github](https://github.com/MargaretKrutikova/reason-react-reductive-example/tree/hooks). It is a small app that has three components subscribed to different portions of state and they only re-render when their state changes.
I also just deployed it for fun on [https://reductive-hooks-example.netlify.com/](https://reductive-hooks-example.netlify.com/), you can check updates in react dev tools and see that stuff re-renders properly.

**Comment by tizoc** [comment]
@MargaretKrutikova that looks great, thank you!

**Comment by MargaretKrutikova** [comment]
No problem! 
There are a few edge cases that need to be handled though, like if `selector ` function changes there should be `useEffect` handling it.
Then there is this "zombie-child" problem that can only appear if children connect to store before their parent and might start querying data that doesn't exist before the parent has time to rerender the children ... weird stuff 🙂

I am thinking about making an initial PR with hooks implementation in `reductive` and see how that goes.

**Comment by tizoc** [comment]
Maybe it is because I come to Reason from OCaml (can't meaningfully compare functions), but I wouldn't worry about changing selectors, I would just mention that the selector should not change in the documentation.

**Comment by tizoc** [comment]
Just pushed [a cleaner implementation](https://github.com/tizoc/refractive) of what I described in #45.

Instead of changing Reductive what I did was to implemente lenses and tracked selectors, with an enhancer for the store. Right not the implementation may be a bit simplistic, in particular, paths-as-strings that get compared may not be enough for all cases. Nothing was done yet about using a PPX rewriter to automatically generate lenses and selectors, also, of the lenses for core datatypes only arrays are implemented for now.

- It works with Reductive stores. A module for tracked selectors has to be instantiated, and the enhancer used with the store. A context has to be instantiated too, it provides the `useSelector` and `useDispatch` hooks.
- Subscriptions are made to selectors (the combination of a lense with a path). This subscribes the component to changes on that path using `useSubscription` to handle it.
- Modifications to the state have to be made through selectors too. This leaves a trace of all the paths that have been touched during the call to the reduce function.
- Only notifications that match the changed paths are fired, which I believe is optimal.

There is a small example in that repository if anyone wants to give it a try.
